### PR TITLE
Binding for octoller (www.octoller.com) V1.0.0 / Try 2

### PR DESCRIPTION
**Pull request by JPlenert**
New binding for octoller. octoller is an open source, smart home and programmable logic controller(www.octoller.com). Right now the binding needs an .Net gateway (octoller.gateway). In a future version it may communicate directly to the octoller.

### CODE REVIEW

File: bundles/binding/org.openhab.binding.octoller/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Manifest-Version: 1.0
+Private-Package: org.openhab.binding.octoller.internal
+Ignore-Package: org.openhab.binding.octoller.internal
+Bundle-License: http://www.eclipse.org/legal/epl-v10.html
+Bundle-Name: openHAB octoller Binding
+Bundle-SymbolicName: org.openhab.binding.octoller
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.0.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/META-INF/MANIFEST.MF]: 1.8.0

File: bundles/binding/org.openhab.binding.octoller/OSGI-INF/binding.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2010-2014, openHAB.org and others.
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/OSGI-INF/binding.xml]: -2015

File: bundles/binding/org.openhab.binding.octoller/OSGI-INF/genericbindingprovider.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2010-2014, openHAB.org and others.
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/OSGI-INF/genericbindingprovider.xml]: -2015

File: bundles/binding/org.openhab.binding.octoller/pom.xml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<parent>
+		<groupId>org.openhab.bundles</groupId>
+		<artifactId>binding</artifactId>
+		<version>1.5.1-SNAPSHOT</version>
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/pom.xml]: 1.8.0-SNAPSHOT

File: bundles/binding/org.openhab.binding.octoller/pom.xml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>org.vafer</groupId>
+				<artifactId>jdeb</artifactId>
+			</plugin>
+		</plugins>
+	</build>
+
+	<url>www.octoller.com</url>
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/pom.xml]: remove

File: bundles/binding/org.openhab.binding.octoller/pom.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	<description>openHAP bnding for octoller (www.octoller.com).
+For use with octoller-Gateway</description>
+	<organization>
+		<name>Joerg Plenert</name>
+	</organization>
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/pom.xml]: remove

File: bundles/binding/org.openhab.binding.octoller/pom.xml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	<version>1.0</version>
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/pom.xml]: remove

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+/**
+ ********************************************************************************************************
+ * openHAB binding for octoller (www.octoller.com)
+ * For use with octoller-Gateway
+ ********************************************************************************************************
+ */
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java]: remove this comment block

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+import org.openhab.binding.octoller.internal.octollerBindingConfig;
+import org.openhab.core.events.EventPublisher;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.PercentType;
+import org.openhab.core.types.State;
+
+/**
+ * @author JPlenert
+ * @since 1.5.1
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java]: 1.8.0

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+ * openHAB binding for octoller (www.octoller.com)
+ * For use with octoller-Gateway
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java]: move comment above @author/@version tag according to the java coding conventions (see https://github.com/openhab/openhab/wiki/How-To-Contribute)

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+		{
+			sb.append("BlockName=");
+			sb.append(config.BlockName);
+			sb.append("|");
+		}
+		
+		return sb.toString();
+	}
+	
+}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/octoller/devicecom/Connection.java]: add newline

**Comment by teichsta** [comment]
Hi @JPlenert,

sorry for letting you waiting soooo long before starting the review. While going through the code i am wondering whether you've read the [How To Contribute](https://github.com/openhab/openhab/wiki/How-To-Contribute) guide. While continuing the code review could you please:

* apply the code formatter (right click > source > format) to all of your files
* update version to 1.8.0
* update the license header
* add a description to all of the classed and remove additional comments below @author/@since

Thanks in advance,

Thomas E.-E.

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+				try
+				{
+					Connection con = new Connection(config.GatewayHost);
+					String result = con.DoCommand(con.BuildCommandString(config, "", ""));
+					con.ProcessResultToPublisher(eventPublisher, itemName, result);
+					con.Close();
+					logger.debug(itemName + " returned " + result);
+				}
+				catch (Exception e) {
+					e.printStackTrace();
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java]: replace with proper logging

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+		logger.debug("execute() method is called!");
+		for (octollerBindingProvider provider : providers) 
+		{
+			for (String itemName : provider.getItemNames()) 
+			{
+				octollerBindingConfig config = (octollerBindingConfig) provider.getConfig(itemName);
+				if (config == null)
+					continue;
+					
+				//eventPublisher.postUpdate(itemName, (State)new PercentType(100));
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java]: remove commented code

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+	/**
+	 * @{inheritDoc}
+	 */
+	@Override
+	protected void internalReceiveUpdate(String itemName, State newState) {
+		// the code being executed when a state was sent on the openHAB
+		// event bus goes here. This method is only called if one of the 
+		// BindingProviders provide a binding for the given 'itemName'.
+		logger.debug("internalReceiveCommand() is called!");
+	}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java]: remove overridden method

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+		if (config != null) {
+			
+			// to override the default refresh interval one has to add a 
+			// parameter to openhab.cfg like <bindingName>:refresh=<intervalInMs>
+			String refreshIntervalString = (String) config.get("refresh");
+			if (StringUtils.isNotBlank(refreshIntervalString)) {
+				refreshInterval = Long.parseLong(refreshIntervalString);
+			}
+			
+			// read further config parameters here ...
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBinding.java]: remove generic comment

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBindingConfig.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */ 
+ 
+package org.openhab.binding.octoller.internal;
+
+import org.openhab.core.binding.BindingConfig;
+
+/**
+ * @author JPlenert
+ * @since 1.5.1
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBindingConfig.java]: 1.8.0

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBindingConfig.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */ 
+ 
+package org.openhab.binding.octoller.internal;
+
+import org.openhab.core.binding.BindingConfig;
+
+/**
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBindingConfig.java]: briefly explain the purpose of this class

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBindingConfig.java [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+ * @author JPlenert
+ * @since 1.5.1
+ * For openHAB binding for octoller (www.octoller.com)
+ */
+public class octollerBindingConfig implements BindingConfig {
+	public String DeviceHost;
+	public String GatewayHost;
+	public String BlockName;
+	public int BlockID;
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerBindingConfig.java]: explain what the members are about

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerGenericBindingProvider.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+import org.openhab.core.items.Item;
+import org.openhab.core.library.items.DimmerItem;
+import org.openhab.core.library.items.SwitchItem;
+import org.openhab.model.item.binding.AbstractGenericBindingProvider;
+import org.openhab.model.item.binding.BindingConfigParseException;
+
+
+/**
+ * This class is responsible for parsing the binding configuration.
+ * For openHAB binding for octoller (www.octoller.com)
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerGenericBindingProvider.java]: remove generic comment and briefly explain what this class is about, also add some Binding configuration examples here.

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerGenericBindingProvider.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+ *  
+ * @author JPlenert
+ * @since 1.5.1
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/internal/octollerGenericBindingProvider.java]: 1.8.0

File: bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/octollerBindingProvider.java [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.octoller;
+
+import org.openhab.core.binding.BindingProvider;
+import org.openhab.binding.octoller.internal.*;
+
+/**
+ * @author JPlenert
+ * @since 1.5.1
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.octoller/src/main/java/org/openhab/binding/octoller/octollerBindingProvider.java]: 1.8.0

**Comment by teichsta** [comment]
i am done with the review. Please find most of the comments inline in addition to the general comments above.

Due to our [1.8 schedule](https://community.openhab.org/t/upcoming-1-8-release/4892/1) i would kindly like to ask you if you would be able to incorporate my feedback until December 30th? Sorry for this hustle as i was the one letting you waiting so long :-).

Thanks in advance, Thomas E.-E.

**Comment by JPlenert** [comment]
Hi Thomas,
I just made the changes. Unfortunately, I’m not able to compile any newer version of openhab. 
After redoing the whole process with maven there are still about 1000 compile errors and missing files even without compiling. Yes, I also changed the Compiler compliance level to 1.7 (and 1.6) and changed the JRE. Maybe this is a problem on Windows 10? With 1.5 on Windows 7 I had no problems.
Hopefully that will change with the final release of 1.8…
However; I had no chance to do any tests!

Regards Joerg

P.S. Have I ever said before said I love C/C++/C#? :-)

Some more details on the problem:

Maven process was OK:

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 29:06 min
[INFO] Finished at: 2015-12-28T12:06:00+01:00
[INFO] Final Memory: 435M/918M
[INFO] ------------------------------------------------------------------------

Errors in Eclipse without compile:
Description	Resource	Path	Location	Type
Project 'org.openhab.binding.urtsi' is missing required source folder: 'src/main/xtend-gen'	org.openhab.binding.urtsi		Build path	Build Path Problem
Project 'org.openhab.model.persistence.ui' is missing required source folder: 'src-gen'	org.openhab.model.persistence.ui		Build path	Build Path Problem
Project 'org.openhab.persistence.jdbc' is missing required library: 'lib/commons-dbutils-1.6.jar'	org.openhab.persistence.jdbc		Build path	Build Path Problem
Project 'org.openhab.persistence.jdbc' is missing required library: 'lib/HikariCP-2.4.2.jar'	org.openhab.persistence.jdbc		Build path	Build Path Problem
Project 'org.openhab.persistence.jdbc' is missing required library: 'lib/yank-3.1.0.jar'	org.openhab.persistence.jdbc		Build path	Build Path Problem
Project 'org.openhab.persistence.jpa' is missing required library: 'lib/derbyclient.jar'	org.openhab.persistence.jpa		Build path	Build Path Problem
Project 'org.openhab.persistence.jpa' is missing required library: 'lib/mysql-connector-java-5.1.26-bin.jar'	org.openhab.persistence.jpa		Build path	Build Path Problem
Project 'org.openhab.persistence.jpa' is missing required library: 'lib/openjpa-all-2.3.0.jar'	org.openhab.persistence.jpa		Build path	Build Path Problem
Project 'org.openhab.persistence.jpa' is missing required library: 'lib/postgresql-9.2-1004.jdbc41.jar'	org.openhab.persistence.jpa		Build path	Build Path Problem

**Comment by JPlenert** [comment]
I just tried to use the vagrant IDE to give it a try. That doesn't work also, because the SSL certificate of "atlas.hashicorp.com" is no longer valid. My IE & Edge also report a broken certificate. No luck!

==> default: Adding box 'rub-a-dub-dub/openhabdev32' (v0.1.0) for provider: virtualbox
    default: Downloading: https://atlas.hashicorp.com/rub-a-dub-dub/boxes/openhabdev32/versions/0.1.0/providers/virtualb
ox.box
    default: Progress: 100% (Rate: 176/s, Estimated time remaining: --:--:--)
An error occurred while downloading the remote file. The error
message, if any, is reproduced below. Please fix this error and try
again.

SSL certificate problem: certificate has expired

**Comment by teichsta** [comment]
Thanks @JPlenert!

**Pull request merged**